Funcky.Test/Extensions/EnumerableExtensions/CartesianProductTest.cs

```diff
@@ -0,0 +1,48 @@
+﻿using System.Collections.Immutable;
+using System.Linq;
+using Funcky.Extensions;
+using Xunit;
+
+namespace Funcky.Test.Extensions.EnumerableExtensions
+{
+    public class CartesianProductTest
+    {
+        [Fact]
+        public void GivenTwoEmptySetsCartesianProductReturnsAnEmptySet()
+        {
+            var ints = Enumerable.Empty<int>();
+            var strings = Enumerable.Empty<string>();
+
+            Assert.Empty(ints.CartesianProduct(strings));
+        }
+
+        [Fact]
+        public void GivenOneEmptySetCartesianProductReturnsAnEmptySet()
+        {
+            var ints = ImmutableList<int>.Empty.Add(1).Add(2).Add(42);
+            var strings = Enumerable.Empty<string>();
+
+            Assert.Empty(ints.CartesianProduct(strings));
+        }
+
+        [Fact]
+        public void GivenTwoSingleElementSequencesWeGetASinglePairBack()
+        {
+            var ints = ImmutableList<int>.Empty.Add(1);
+            var strings = ImmutableList<string>.Empty.Add("Hello");
+
+            var product = ints.CartesianProduct(strings);
+            Assert.Single(product);
+        }
+
+        [Fact]
+        public void GivenTowSequencesWithMultipleItemsCartesianProductWithProjectReturnsAllPossibleCombinations()
```

Comment:
Add another test with two sequences that aren't the same length

Reply:
addressed in 2168591